File: 150.evaluate-reverse-polish-notation.cpp
/*
 * @lc app=leetcode id=150 lang=cpp
 *
 * [150] Evaluate Reverse Polish Notation
 */

// @lc code=start
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<long long> stk;
        for(auto a:tokens){
            if(isOper(a)){
                long first=stk.top();
                stk.pop();
                long second=stk.top();
                stk.pop();
                if(a=="+"){
                    stk.push(first+second);
                }
                else if(a=="-"){
                    stk.push(second-first);
                }
                else if (a=="*"){
                    stk.push((long long)first*second);
                }
                else if(a=="/"){
                    stk.push((second/first));
                }
            }
            else{
                stk.push(stoi(a));
            }
        }
        return stk.top();
    }
    bool isOper(string c){
        if(c=="+" || c=="-"|| c=="/"|| c=="*"){
            return true;
        }
        return false;
    }
};

// @lc code=end

File: 150.evaluate-reverse-polish-notation.py
#
# @lc app=leetcode id=150 lang=python3
#
# [150] Evaluate Reverse Polish Notation
#

# @lc code=start
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stk = deque()

        for token in tokens:
            if token in '+-*/':
                val1 = stk.pop()
                val2 = stk.pop()
                ans = 0
                if token == '+':
                    ans = (val1 + val2)
                elif token == '-':
                    ans =(val2 - val1)
                elif token == '*':
                    ans =(val1 * val2)
                else:
                    ans =(int(val2 / val1))
                stk.append(ans)
            else:
                stk.append(int(token))
        
        return stk.pop()
# @lc code=end

